src/compiler/llvm/mir.rs

```diff
@@ -440,15 +440,17 @@ impl<'p, 'module, 'ctx>
     fn add_type(&mut self, id: TypeId, ty: &MirTypeDef) -> Result<(), TransformerError> {
         debug!("Adding a type to the Module");
 
-        let previous_value = ty
-            .into_basic_type_enum(self)
-            .and_then(|llvm_ty| self.ty_table.insert(id, llvm_ty));
-
-        // If `insert` returns `None` then it means there was no previous value associated with `id`
-        // otherwise, `id` was already defined and this should thrown an error.
-        match previous_value {
-            None => Ok(()),
-            Some(_) => Err(TransformerError::TypeAlreadyDefined),
+        // If type is already in the table then skip
```

Comment:
The semantics on this are weird: do I have an invariant in terms of insuring that this `add_type` will only be called on a type T if all the constituent types have already been passed to `add_type`? Is that an assumption that can be conveyed or expected via the API presented here?

Given that, what are the errors that could be expect to be thrown from this function?

Reply:
What I can do is build a partial order of the TypeIds based upon insuring that a type T is in the list before every type that contains T.  If that invariant is given, then this function can assume that any attempt to re-add T is a violation of expectations and something wrong has happened.

Reply:
There is the problem of cycles, for example, the following code is legal:

```
struct S {
   ptr: *const S,
}
```